### ISSUE TITLE
[BUG] EventProcessorClient fails to recover after AMQP link loss

### ISSUE THREAD

**Issue by pvorb** [Azure/azure-sdk-for-java#23793]
**Describe the bug**
After around 17 hours of successful consumption, our `EventProcessorClient` suddenly closed its AMQP link and tried to request another one. That was unsuccessful, but I can't tell why. I would have expected the `EventProcessorClient` to recover and start consuming messages again.

***Exception or Stack Trace***
There was no exception at all. You can follow the events in this log file: [2021-08-27_amqpSessionClosed.log](https://github.com/Azure/azure-sdk-for-java/files/7067017/2021-08-27_amqpSessionClosed.log)

**To Reproduce**
* Create an Event Hub with one partition
* Configure an `EventProcessorClient` that consumes from this Event Hub:
  * Batch size: 500
  * Max wait time: 5 seconds
  * Load balancing strategy: BALANCED
  * Offset reset position: earliest
  * Blob checkpoint store
  * Use the default retry settings
* Produce many messages (we successfully received ~1.5 million before the AMQP link was closed) and wait for several hours :wink:

***Code Snippet***

I hope you don't mind that this is in Kotlin:

~~~ kotlin
val client = EventProcessorClientBuilder()
		.credential(eventHubHostname, eventHubName, getSharedAccessKeyCredential())
		.transportType(transportType)
		.proxyOptions(getProxyOptions())
		.customEndpointAddress(customEndpointAddress)
		.clientOptions(getClientOptions())
		.initialPartitionEventPosition(initialPartitionEventPositions())
		.consumerGroup(consumerGroup)
		.loadBalancingStrategy(loadBalancingStrategy)
		.loadBalancingUpdateInterval(loadBalancingUpdateInterval)
		.partitionOwnershipExpirationInterval(partitionOwnershipExpirationInterval)
		.checkpointStore(BlobCheckpointStore(blobCheckpointStore.createAsyncClient()))
		.processPartitionInitialization {
			logger.debug { "Initialized consumption of ${it.partitionContext}" }
		}
		.processPartitionClose {
			logger.info { "Closed consumption of ${it.partitionContext}" }
		}
		.processError {
			logger.warn(it.throwable) { "Failed to consume from ${it.partitionContext}'" }
		}
		.processEventBatch(
			batchConsumer,
			properties.maxBatchSize,
			properties.maxWaitTime
		)
		.trackLastEnqueuedEventProperties(true)
		.buildEventProcessorClient()

client.start()
~~~

This doesn't directly cause the problem. The code is functional from as far as I can tell, but it caused the problem that I described above.

**Expected behavior**
I'd expect consumption to recover from connection problems without manual intervention.

**Screenshots**
![grafik](https://user-images.githubusercontent.com/183534/131127564-3e7608d1-85d7-4ff6-a2f8-e84242672a9a.png)

In the screenshot you can see in the time range between 6 am and 12 pm that the outgoing messages suddenly drop to zero. They go up again once the service had been restarted and the processor received the messages that accumulated over several hours on the event hub, so the number of outgoing messages is much higher than before.

The short period at the beginning is where we had another consumer running. This is not related to the issue.

**Setup (please complete the following information):**
 - OS: Linux (we're running on Azure Kubernetes Services)
 - IDE: IntelliJ
 - Library/Libraries: e.g. com.azure:azure-messaging-eventhubs:5.9.0, com.azure:azure-messaging-eventhubs-checkpointstore-blob:1.8.1 and all their transitive dependencies
 - Java version: 11 (official AdoptOpenJDK Docker image)
 - App Server/Environment: Tomcat embedded (via Spring Boot)
 - Frameworks: Spring Boot
  
**Additional context**
There's no more context I can give.

**Information Checklist**
- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

**Comment by anuchandy** [comment]
@pvorb Thanks for sharing the log:

```
namespace[our-event-hubs-namespace.servicebus.windows.net] entityPath[ingest]: Next AMQP channel received, updating 0 current subscribers
```

The above line explains that a new connection is created after the previous one is disposed, but there seems to be no downstream to use that connection (0 subscribers).  Are there any more log entries from the SDK after this point? If so, could you share?

**Comment by pvorb** [comment]
@anuchandy Thanks for having a look.

This is the full log that was still available at the time before we restarted the service: [2021-08-27_amqpSessionClosed_full.log.gz](https://github.com/Azure/azure-sdk-for-java/files/7101278/2021-08-27_amqpSessionClosed_full.log.gz)

The line you mentioned is at line 41413, so there's also a lot of logs before the issue occurred. I hope you can make some sense from it. Shortly after the mentioned line, the Reactor seems to receive a shutdown signal and after that point, only the `PartitionBasedLoadBalancer` keeps running every 10 seconds. :shrug:

**Comment by conniey** [comment]
Fixed in #24141.
